File: LeetCode/iter1/java/743.network-delay-time.java
package org.ironman.base.playground.leetcode;

import java.util.LinkedList;
import java.util.Queue;

public class Solution {

    public static void main(String[] args) {
        testCase1();
        System.out.println();
    }

    private static void testCase1() {
        int[] edge1 = new int[]{2, 1, 1};
        int[] edge2 = new int[]{2, 3, 1};
        int[] edge3 = new int[]{3, 4, 1};
        int[][] edges = new int[][]{edge1, edge2, edge3};

        System.out.println(networkDelayTime(edges, 4, 2));
    }

    public static int networkDelayTime(int[][] times, int N, int K) {
        int[] dist = new int[N + 1];
        for (int i = 0; i < N + 1; i++) {
            dist[i] = Integer.MAX_VALUE;
        }
        //distance of source node is 0
        dist[K] = 0;
        //traverse N node
        for (int i = 0; i < N; i++) {
            for (int[] edge : times) {
                int s = edge[0], t = edge[1], cost = edge[2];
                if (dist[s] != Integer.MAX_VALUE && dist[t] > dist[s] + cost) {
                    dist[t] = dist[s] + cost;
                }
            }
        }

        int maxCost = Integer.MIN_VALUE;
        for (int i = 1; i <= N; i++) {
            maxCost = Math.max(maxCost, dist[i]);
        }
        return maxCost == Integer.MAX_VALUE ? -1 : maxCost;
    }

    private static void printLevelTree(TreeNode root) {
        Queue<Pair<TreeNode, Integer>> q = new LinkedList<>();
        q.add(new Pair<>(root, 1));
        int curLv = 1;
        while (!q.isEmpty()) {
            Pair<TreeNode, Integer> tmpPair = q.poll();
            if (tmpPair.getValue() != curLv) {
                curLv = tmpPair.getValue();
                System.out.println();
            }

            System.out.print(tmpPair.getKey().val + " ");
            if (tmpPair.getKey().left != null) {
                q.add(new Pair<>(tmpPair.getKey().left, tmpPair.getValue() + 1));
            }

            if (tmpPair.getKey().right != null) {
                q.add(new Pair<>(tmpPair.getKey().right, tmpPair.getValue() + 1));
            }
        }
        System.out.println();
    }

    static public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    static public class Pair<K, V> {
        private K key;
        private V value;

        Pair(K _key, V _value) {
            key = _key;
            value = _value;
        }

        public K getKey() {
            return key;
        }

        public V getValue() {
            return value;
        }
    }
}

File: LeetCode/iter1/c++/743.network-delay-time.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

struct TreeNode {
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int networkDelayTime(vector<vector<int>>& times, int N, int K) {
		vector<int> dist(N + 1, INT_MAX);
		dist[K] = 0;
		for (int i = 0; i < N; i++) {
			for (vector<int> edge : times) {
				int s = edge[0], t = edge[1], cost = edge[2];
				if (dist[s] != INT_MAX && dist[t] > dist[s] + cost) {
					dist[t] = dist[s] + cost;
				}
			}
		}

		int maxVal = INT_MIN;
		for (int i = 1; i <= N; i++) {
			if (dist[i] > maxVal) {
				maxVal = dist[i];
			}
		}

		return maxVal == INT_MAX ? -1 : maxVal;
	}
};

void testPrint(TreeNode* treeNode) {
	queue<pair<TreeNode*, int>> treeQ;
	treeQ.push(pair<TreeNode*, int>(treeNode, 1));
	int curLv = 1;

	while (!treeQ.empty()) {
		pair<TreeNode*, int>tmpPair = treeQ.front();
		treeQ.pop();

		if (curLv != tmpPair.second) {
			curLv = tmpPair.second;
			cout << endl;
		}

		cout << tmpPair.first->val << " ";
		if (tmpPair.first->left != NULL) {
			treeQ.push(pair<TreeNode*, int>(tmpPair.first->left, tmpPair.second + 1));
		}
		if (tmpPair.first->right != NULL) {
			treeQ.push(pair<TreeNode*, int>(tmpPair.first->right, tmpPair.second + 1));
		}
	}
}

void testCase1() {
	vector<int>v1;
	v1.push_back(2);
	v1.push_back(1);
	v1.push_back(1);

	vector<int>v2;
	v2.push_back(2);
	v2.push_back(3);
	v2.push_back(1);

	vector<int>v3;
	v3.push_back(3);
	v3.push_back(4);
	v3.push_back(1);

	vector<vector<int>> v;
	v.push_back(v1);
	v.push_back(v2);
	v.push_back(v3);

	Solution s;
	cout << s.networkDelayTime(v, 4, 2) << endl;
}

int main(void)
{
	testCase1();
	cout << endl;

	return 0;
}
